### PR TITLE
Texturepacker: add compressed texture support

### PR DESCRIPTION
**Pull request by sarbes**
## Description
The PR adds support for compressed textures to the texture packer. Compressed textures can be supplied via KTX containers. 

There are switches to allow the packer to substitute normally uncompressed files with compressed counterparts. This makes it possible to generate texture packs which are tailored to different families of devices. For more information, see https://github.com/xbmc/xbmc/blob/31b17010d14b78268241f311de24149eccf127d7/tools/depends/native/TexturePacker/src/TexturePacker.cpp#L127 

## Motivation and context
The texture footprint of Estuary in Omega was 47.6MB. By using single and dual channel formats, it had been reduced to 12.5MB. With most compression schemes, it can be reduced further to around 6.5MB. By using ASTC, the footprint is just 3.1MB, a 93% reduction compared to Omega.

By using compressed formats, the texture upload should be faster, as less data is read from disk and moved around. The amount of GPU-RAM used should be considerably lower. The same should apply for memory bandwidth used by the GPU. This frees precious resources for the rest of the system.

The loss in image quality is virtually imperceivable. 

To generate the KTX textures, I've created a python script which feeds various encoders. It is pretty hacky atm, but I'm planning on submitting it in the future.

The next step would be to add a way to distribute and enable such texture packs. For testing, I'm attaching a XBT file for Estuary, containing BPTC and RGTC textures. By replacing the file "addons/skin.estuary/media/Textures.xbt" with [Textures.xbt.zip](https://github.com/user-attachments/files/17422875/Textures.xbt.zip), the skin should load the compressed textures on Linux/Mesa systems.

## How has this been tested?
Various texture profiles render fine on my desktop system.

## What is the effect on users?
Right now, it won't have any effect.

## Screenshots (if appropriate):
All XBT textures in this screenshot are in an ASTC 8x8 format, with a bitrate of 2 bit/pixel. 
image

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

File: tools/depends/native/TexturePacker/src/TexturePacker.cpp [hunk 1]
```diff
@@ -185,13 +264,65 @@ void TexturePacker::CreateSkeletonHeader(CXBTFWriter& xbtfWriter,
   }
 }
 
+void TexturePacker::SubstitudeFile(CXBTFWriter& xbtfWriter, const std::string& fileName)
```
Comment (sarbes) [inline tools/depends/native/TexturePacker/src/TexturePacker.cpp:267]: Eh, spelling is hard. I'll correct this and other instances later.

**Comment by neo1973** [comment]
Works on my AMD Ryzen 7 5700U but not on my Raspberry Pi 3B:

IMG_20241018_012832
IMG_20241018_012854

Maybe a hardware limitation?

Debug log: https://paste.kodi.tv/ugokotuxoy.kodi

File: tools/depends/native/TexturePacker/src/TexturePacker.cpp (hunk 1 above)
Comment (garbear) [reply tools/depends/native/TexturePacker/src/TexturePacker.cpp:267]: We can't all be excellent at speling.

**Comment by sarbes** [comment]
Interesting, the Pi 3 (VideoCore IV) exposes all but BPTC/RGTC. In reality, it just supports ETC1, all other formats are emulated. Mesa just decodes/transcodes in this case.

Such textures make heavy use of component swizzling, which is not available on GLES 2.0. This is also a problem for uncompressed alpha only textures (those make up ~95% of all of Estuary's textures), which we currently convert on the CPU.

In short, ETC1 will be the only viable format on GLES 2.0, and it needs some specialized shaders.

**Comment by sarbes** [comment]
The Pi 4/5 (Videocore VI/VII) support ASTC and swizzling. If you have one, you could give it a shot: 
[Textures.xbt.zip](https://github.com/user-attachments/files/17431243/Textures.xbt.zip)

I've uploaded a branch which contains the compressed textures: https://github.com/sarbes/xbmc/tree/estuary-compression-test. Here is the command for generating an XBT with ASTC files: `./path/to/build/texturepacker/TexturePacker -input /path/to/other/branch/media/ -output Textures.xbt -dupecheck -astc`

**Comment by zag2me** [comment]
How will the effect for example someone updating a texture in Estuary?

**Comment by sarbes** [comment]
Errr. Could you rephrase that?

**Comment by sarbes** [comment]
With https://github.com/xbmc/xbmc/pull/25855 in addition to this PR, GLES 2.0 systems should be able to display ETC1 textures with 111R swizzles - so 90% of the textures in the branch above.

**Comment by neo1973** [comment]
> With https://github.com/xbmc/xbmc/pull/25855 in addition to this PR, GLES 2.0 systems should be able to display ETC1 textures with 111R swizzles - so 90% of the textures in the branch above.

I can confirm that together with the Textures.xtb from https://github.com/xbmc/xbmc/pull/25846#issuecomment-2421922206 everything looks correct on the RPi3B.

In addition I tested this PR alone together with the initial Textures.xtb on a RK3568 based SBC (the one from Conrad on sale you mentioned :wink:) and it was working correctly as well.

I'll make a review of the code next.

**Comment by zag2me** [comment]
> Errr. Could you rephrase that?

Was just wondering if this will make it harder for people to change textures in Estuary. Or does texture packer handle that?

**Comment by sarbes** [comment]
> > Errr. Could you rephrase that?
> 
> Was just wondering if this will make it harder for people to change textures in Estuary. Or does texture packer handle that?

This PR is just about adding the ability to read KTX files to the texture packer and optionally substituting textures. How those textures are supplied and how the resulting XBT is handled is not part of it.

File: tools/depends/native/TexturePacker/src/decoder/KTXDecoder.cpp [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+  {"KD_TEX_SWIZ_RGB_SDF", KD_TEX_SWIZ_RGB_SDF},
+  {"KD_TEX_SWIZ_MSDF", KD_TEX_SWIZ_MSDF},
+};
+
+static const std::map<std::string, KD_TEX_ALPHA> AlphaKeys
+{
+  {"KD_TEX_ALPHA_STRAIGHT", KD_TEX_ALPHA_STRAIGHT},
+  {"KD_TEX_ALPHA_OPAQUE", KD_TEX_ALPHA_OPAQUE},
+  {"KD_TEX_ALPHA_PREMULTIPLIED", KD_TEX_ALPHA_PREMULTIPLIED},
+};
```
[the remark this answers is not in the document]
Comment (sarbes) [reply tools/depends/native/TexturePacker/src/decoder/KTXDecoder.cpp:109]: I really don't want to introduce more dependencies on core code. I would hope that we could move the packer to its own repository someday (maybe that's a nice devcon project?).

File: tools/depends/native/TexturePacker/src/decoder/KTXDecoder.cpp [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+
+  std::array<unsigned char, 4> endianess;
+  if (fp.Read(&endianess, sizeof(endianess)) != sizeof(endianess))
+    return false;
+
+  if (endianess != m_endianess)
+    return false;
+
+  return true;
+}
```
[the remark this answers is not in the document]
Comment (sarbes) [reply tools/depends/native/TexturePacker/src/decoder/KTXDecoder.cpp:145]: Sure, but does it matter for such an auxiliary tool?

[the remark this answers is not in the document]
Comment (neo1973) [reply tools/depends/native/TexturePacker/src/decoder/KTXDecoder.cpp:145]: In terms of performance? No. Is it more readable? I'd say yes, but of course that's a matter of taste. I'm fine either way and won't oppose this PR because of this.

File: tools/depends/native/TexturePacker/src/decoder/KTXDecoder.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (neo1973) [reply tools/depends/native/TexturePacker/src/decoder/KTXDecoder.cpp:109]: Then lets make it a `std::unordered_map`.

> I would hope that we could move the packer to its own repository someday (maybe that's a nice devcon project?).

I'm a monorepo guy, but I guess that's another discussion :wink:

**Comment by garbear** [comment]
Is this wanted for Alpha 1?

**Comment by garbear** [comment]
Is this wanted for Alpha 2, due this weekend?

**Comment by garbear** [comment]
Bumped to Alpha 3

**Comment by garbear** [comment]
Is this wanted for Beta 2? Or do we bump to v23?
